networkx/algorithms/euler.py

```diff
@@ -147,3 +157,154 @@ def eulerian_circuit(G, source=None):
             arbitrary_edge = next(edges(current_vertex))
             vertex_stack.append(get_vertex(arbitrary_edge))
             g.remove_edge(*arbitrary_edge)
+
+
+def eulerian_path(G):
+    """Returns the edges of an Eulerian path in G, (if it exits).
+
+    An Eulerian path is a path that crosses every edge in G
+    exactly once.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    edges: generator
+        A generator that produces edges in the Eulerian path.
+
+    Raises
+    ------
+    NetworkXError: If the graph does not have an Eulerian path.
+
+    Notes
+    -----
+    Linear time algorithm, adapted from [1]_ and [3]_.
+    Information about euler paths in [2]_.
+    Code for Eulerian circuit in [3]_.
+
+    Important: In [1], euler path is in reverse order,
+    this implementation gives the path in correct order
+    as in [3]_ for eulerian_circuit. The distinction for
+    directed graph is in using the in_degree neighbors, not the out
+    ones. for undirected, it is using itemgetter(1) for get_vertex,
+    which uses the correct vertex for this order. Also, every graph
+    has an even number of odd vertices by the Handshaking Theorem [4]_.
+
+    References
+    ----------
+    .. [1] http://www.graph-magics.com/articles/euler.php
+    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
+    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
+    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
+
+    Examples
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')])
+    >>> list(nx.find_eulerian_path(G))
+    [('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')]
+
+    >>> G = nx.Digraph([(1, 2), (2, 3)])
+    >>> list(nx.find_eulerian_path(G))
+    [(1,2),(2,3)]
+    """
+    from operator import itemgetter
+
+    # Verify that graph is connected, short circuit
+    if G.is_directed() and not nx.is_weakly_connected(G):
+        raise nx.NetworkXError("G is not connected.")
+
+    # is undirected
+    if not G.is_directed() and not nx.is_connected(G):
+        raise nx.NetworkXError("G is not connected.")
+
+    # Now verify if has an Eulerian circuit: even condition of all nodes is satified.
+    if nx.is_eulerian(G):
+        # if we have an Eulerian cycle, generate its edges
+        for edge in nx.eulerian_circuit(G):
+            yield edge
+        # and we're done
+        return
+    
+    # Not all vertex have even degree, check if exactly two vertex have odd degrees.
+    # If yes, then there is an Euler path. If not, raise an error (no euler path can be found)
+    g = G.__class__(G)  # copy graph structure (not attributes)
+
+    # list to check the odd degree condition, and a flag
+    check_odd = []
+    directed = g.is_directed()
```

Comment:
This should be moved to the beginning of the function, so it can be used in the checks for connectedness. Also, it should be called `is_directed`, since it is a Boolean.